ci/pkl-pipelines/pay-dev/e2e-helpers.pkl

```diff
@@ -0,0 +1,707 @@
+  ) {
+    source { ["tag_regex"] = "reverse_proxy_alpha_release-(.*)" }
+  }
+
+  (shared_resources.payECRResource(
+    "endtoend-ecr-registry-test", "govukpay/endtoend", "pay_aws_test_account_id")
+  ) {
+    source { ["tag"] = "latest" }
+  }
+
+  (shared_resources.payECRResourceWithVariant(
+    "endtoend-candidate-ecr-registry-test",
+    "govukpay/endtoend",
+    "pay_aws_test_account_id",
+    "candidate"))
+
+  (shared_resources.payECRResourceWithVariant(
+    "reverse-proxy-candidate-ecr-registry-test",
+    "govukpay/reverse-proxy",
+    "pay_aws_test_account_id",
+    "candidate"))
+
+  (shared_resources.payECRResourceWithVariant(
+    "stubs-candidate-ecr-registry-test",
+    "govukpay/stubs",
+    "pay_aws_test_account_id",
+    "candidate"))
+
+  shared_resources.payDockerHubResource(
+    "endtoend-dockerhub", 
+    "governmentdigitalservice/pay-endtoend", 
+    "latest-master")
+
+  (shared_resources.payDockerHubResource("postgres-15-alpine", "postgres", "15-alpine")) {
+    check_every = "1h" 
+  }
+
+  (shared_resources.payECRResource(
+    "ecr-postgres-15-alpine", "postgres", "pay_aws_test_account_id")
+  ) {
+    check_every = "never"
+    source { ["tag"] = "15-alpine" }
+  }
+
+  (shared_resources.payDockerHubResource(
+    "localstack-localstack-3", 
+    "localstack/localstack", "3")
+  ) {
+    check_every = "1h"
+  }
+
+  (shared_resources.payECRResource(
+    "ecr-localstack-localstack-3", 
+    "localstack/localstack", 
+    "pay_aws_test_account_id")
+  ) {
+    check_every = "never"
+    source { ["tag"] = "3" }
+  }
+
+  (shared_resources.payDockerHubResource(
+    "selenium-standalone-chrome-3-141-59",
+    "selenium/standalone-chrome",
+     "3.141.59")) {
+    check_every = "1h"
+  }
+
+  (shared_resources.payECRResource(
+    "ecr-selenium-standalone-chrome-3-141-59",
+    "selenium/standalone-chrome",
+    "pay_aws_test_account_id")
+  ) {
+    check_every = "never"
+    source { ["tag"] = "3.141.59" }
+  }
+
+  shared_resources.slackNotificationResource
+}
+
+resource_types = new {
+  shared_resources.slackNotificationResourceType
+}
+
+jobs = new {
+  pipeline_self_update.PayPipelineSelfUpdateJob("pay-dev/e2e-helpers.pkl")
+
+  new {
+    name = "build-and-push-endtoend-candidate"
+    plan {
+      new InParallelStep {
+        in_parallel = new Listing<Step> {
+          new GetStep { get = "endtoend-git-release" trigger = true }
+          new GetStep { get = "pay-ci" }
+        }
+      }
+
+      shared_resources.generateDockerCredsConfigStep
+      parseReleaseTag("endtoend-git-release")
+      new LoadVarStep { load_var = "release_number_tag" file = "tags/release-number" }
+
+      new InParallelStep {  // One step in parallel, but that's what the original does
+        in_parallel = new Listing<Step> {
+          buildImage("build-endtoend-image", "endtoend-git-release")
+        }
+      }
+
+      putCandidateImage("endtoend-candidate-ecr-registry-test")
```

Comment:
This can be replaced with below? If not `putCandidateImage` function can use putStep 
```suggestion
      putStep("endtoend-candidate-ecr-registry-test", "image/image.tar, "tags/candidate-tag", true)
```